lib/web_ui/lib/src/engine/pointer_converter.dart

```diff
@@ -0,0 +1,255 @@
+      obscured: datum.obscured,
+      synthesized: true,
+      pressure: datum.pressure,
+      pressureMin: datum.pressureMin,
+      pressureMax: datum.pressureMax,
+      distance: datum.distance,
+      distanceMax: datum.distanceMax,
+      size: datum.size,
+      radiusMajor: datum.radiusMajor,
+      radiusMinor: datum.radiusMinor,
+      radiusMin: datum.radiusMin,
+      radiusMax: datum.radiusMax,
+      orientation: datum.orientation,
+      tilt: datum.tilt,
+      platformData: datum.platformData,
+      scrollDeltaX: datum.scrollDeltaX,
+      scrollDeltaY: datum.scrollDeltaY,
+    );
+  }
+
+  /// Convert the given list pointer data into a sequence of framework-compatible
+  /// pointer data.
+  Iterable<ui.PointerData> convert(List<ui.PointerData> data) sync* {
+    for (ui.PointerData datum in data) {
+      assert(datum.change != null);
+      if (datum.signalKind == null ||
+        datum.signalKind == ui.PointerSignalKind.none) {
+        switch (datum.change) {
+          case ui.PointerChange.add:
+            assert(!_pointers.containsKey(datum.device));
+            _ensureStateForPointer(datum, datum.physicalX, datum.physicalY);
+            assert(!_locationHasChanged(datum));
+            yield _generateCompletePointerData(datum);
+            break;
+          case ui.PointerChange.hover:
+            final bool alreadyAdded = _pointers.containsKey(datum.device);
+            final _PointerState state = _ensureStateForPointer(
+              datum, datum.physicalX, datum.physicalY);
+            assert(!state.down);
+            if (!alreadyAdded) {
+              // Synthesizes an add pointer data.
+              yield _synthesizeFrom(datum, ui.PointerChange.add);
+            }
+            yield _generateCompletePointerData(datum);
+            break;
+          case ui.PointerChange.down:
+            final bool alreadyAdded = _pointers.containsKey(datum.device);
+            final _PointerState state = _ensureStateForPointer(
+              datum, datum.physicalX, datum.physicalY);
+            assert(!state.down);
+            if (!alreadyAdded) {
+              // Synthesizes an add pointer data.
+              yield _synthesizeFrom(datum, ui.PointerChange.add);
+            }
+            assert(!_locationHasChanged(datum));
+            state.startNewPointer();
+            state.down = true;
+            yield _generateCompletePointerData(datum);
+            break;
+          case ui.PointerChange.move:
+            final bool alreadyAdded = _pointers.containsKey(datum.device);
+            final _PointerState state = _ensureStateForPointer(
+              datum, datum.physicalX, datum.physicalY);
+            if (!alreadyAdded) {
+              // Synthesizes an add pointer data and down pointer data.
+              yield _synthesizeFrom(datum, ui.PointerChange.add);
+              yield _synthesizeFrom(datum, ui.PointerChange.down);
+            }
+            assert(state.down);
```

Comment:
I'm probably missing something here, but it seems like this assertion always fails if `alreadyAdded` is false. We could do one of two things:
1. Do `state.down = true;` when we synthesize `PointerChange.down` above.
2. Disallow `PointerChange.move` if the pointer hasn't already been added.

I'm in favor of option 2, but I'm open for discussing it further.

Reply:
oh I missing a state.down=true. I added this part when i was confused at https://github.com/flutter/engine/pull/14082#discussion_r352911298
I should remove this synthesizing all together